sync/azuredatabricks.py

```diff
@@ -0,0 +1,411 @@
+
+        def write_file(body: bytes):
+            logger.info("Saving state to DBFS")
+            write_dbfs_file(path, body, dbx_client)
+
+    resource_group_name = _get_databricks_resource_group_name()
+    if not resource_group_name:
+        logger.warning("Failed to find Databricks managed resource group")
+
+    previous_instances = {}
+
+    while True:
+        try:
+            compute = _get_azure_client(ComputeManagementClient)
+            if resource_group_name:
+                vms = compute.virtual_machines.list(resource_group_name=resource_group_name)
+            else:
+                vms = compute.virtual_machines.list_all()
+            new_instances = [
+                vm.as_dict()
+                for vm in vms
+                if vm.tags.get("Vendor") == "Databricks" and vm.tags.get("ClusterId") == cluster_id
+            ]
+
+            new_instance_id_to_reservation = dict(
+                zip(
+                    (vm["id"] for vm in new_instances),
+                    new_instances,
+                )
+            )
+
+            old_instance_id_to_reservation = dict(
+                zip(
+                    (vm["id"] for vm in previous_instances),
+                    previous_instances,
+                )
+            )
+
+            old_instance_ids = set(old_instance_id_to_reservation)
+            new_instance_ids = set(new_instance_id_to_reservation)
+
+            # If we have the exact same set of instances, prefer the new set...
+            if old_instance_ids == new_instance_ids:
+                instances = new_instances
+            else:
+                # Otherwise, update old references and include any new instances in the list
+                newly_added_instance_ids = new_instance_ids.difference(old_instance_ids)
+                updated_instance_ids = new_instance_ids.intersection(old_instance_ids)
+                removed_instance_ids = old_instance_ids.difference(new_instance_ids)
+
+                removed_instances = [
+                    old_instance_id_to_reservation[id] for id in removed_instance_ids
+                ]
+                updated_instances = [
+                    new_instance_id_to_reservation[id] for id in updated_instance_ids
+                ]
+                new_instances = [
+                    new_instance_id_to_reservation[id] for id in newly_added_instance_ids
+                ]
+
+                instances = [*removed_instances, *updated_instances, *new_instances]
+
+            write_file(orjson.dumps(instances))
+
+            previous_instances = instances
+        except Exception as e:
+            logger.error(f"Exception encountered while polling cluster: {e}")
+
+        sleep(polling_period)
+
+
+def _get_databricks_resource_group_name() -> str:
+    resources = _get_azure_client(ResourceManagementClient)
+
+    for workspace_item in resources.resources.list(
+        filter="resourceType eq 'Microsoft.Databricks/workspaces'"
```

Comment:
love the weird query language, haha. Thanks Microsoft

Reply:
Yeah, and that it's partially implemented on just about every API is really great too. They love to keep you guessing.